File: backend/src/services/database_operations.py
"""
Database and file processing operations.

This module contains operations for:
- File content extraction (PDF, text, Excel)
- Employee UUID lookup from database
- File storage uploads
- Document insertion into database
"""

import logging
from datetime import datetime, timezone
from typing import Dict, Any, Optional, Tuple

from src.core.file_readers import read_pdf, read_text_file, read_excel
from .helpers import get_supabase_client, guess_content_type, make_storage_path, BUCKET

logger = logging.getLogger(__name__)


def read_file(filename: str, file_bytes: bytes) -> Dict[str, Any]:
    """
    Read and extract content from a file based on its extension.
    
    Supports:
    - PDF files (.pdf)
    - Text files (.txt, .md)
    - Excel files (.xlsx, .xlsm, .xls)
    
    Args:
        filename: Name of the file (with extension)
        file_bytes: File content as bytes
    
    Returns:
        Dictionary with:
        - "content_text" (str): Extracted text content
        - "content_structured" (dict, optional): Structured data (for Excel files)
        - "warnings" (list): List of warning messages
    """
    lower = filename.lower()
    
    if lower.endswith(".pdf"):
        response = read_pdf(file_bytes)
    elif lower.endswith(".txt") or lower.endswith(".md"):
        response = read_text_file(file_bytes)
    elif lower.endswith(".xlsx") or lower.endswith(".xlsm") or lower.endswith(".xls"):
        response = read_excel(file_bytes)
    else:
        response = {
            "kind": "unknown",
            "content_text": "",
            "content_structured": None,
            "warnings": ["File extension not recognized"],
        }
    
    return {
        "content_text": response.get("content_text", ""),
        "content_structured": response.get("content_structured", None),
        "warnings": response.get("warnings", []),
    }


def get_employee_uuid(employee_id: str) -> Tuple[Optional[str], Optional[str]]:
    """
    Query the employees table to get the UUID of an employee by their employee_id.
    
    Args:
        employee_id: The employee ID to look up (e.g., "EMP-005")
    
    Returns:
        Tuple of (uuid, error):
        - uuid (str, optional): The UUID of the employee if found
        - error (str, optional): Error message if query failed or employee not found
    """
    try:
        supabase = get_supabase_client()
        
        # Normalize employee_id by removing dashes (backend stores as 'EMP001', frontend may send 'EMP-001')
        normalized_employee_id = employee_id.replace("-", "")
        
        logger.info(f"Querying employees table for employee_id: {employee_id} (normalized: {normalized_employee_id})")
        
        # Query employees table where employee_id matches
        response = supabase.table("employees").select("id").eq("employee_id", normalized_employee_id).execute()
        
        # Check for errors
        if hasattr(response, 'error') and response.error:
            error_msg = f"Database query error: {response.error}"
            logger.error(error_msg)
            return None, error_msg
        
        # Extract UUID from response
        if hasattr(response, 'data') and response.data and len(response.data) > 0:
            employee_uuid = response.data[0].get("id")
            if employee_uuid:
                logger.info(f"Found employee UUID: {employee_uuid} for employee_id: {employee_id}")
                return str(employee_uuid), None
        
        # Employee not found
        error_msg = f"Employee with employee_id '{employee_id}' not found"
        logger.warning(error_msg)
        return None, error_msg
    
    except Exception as e:
        error_msg = f"Failed to query employee UUID: {str(e)}"
        logger.error(error_msg, exc_info=True)
        return None, error_msg


def upload_to_storage(employee_uuid: str, filename: str, file_bytes: bytes) -> Tuple[Optional[str], Optional[str]]:
    """
    Upload a file to a PRIVATE Supabase Storage bucket.
    
    Args:
        employee_uuid: UUID of the employee who owns/uploaded the document
        filename: Original filename of the file being uploaded
        file_bytes: The file content as bytes
    
    Returns:
        Tuple of (file_path, error):
        - file_path (str, optional): The storage path of the uploaded file if successful
        - error (str, optional): Error message if upload failed
    """
    try:
        supabase = get_supabase_client()
        
        # Generate storage path
        path = make_storage_path(employee_uuid, filename)
        
        # Guess content type from filename
        content_type = guess_content_type(filename)
        
        logger.info(f"Uploading file to storage: bucket={BUCKET}, path={path}, content_type={content_type}")
        
        # Upload to Supabase Storage
        response = supabase.storage.from_(BUCKET).upload(
            path=path,
            file=file_bytes,
            file_options={
                "content-type": content_type
            }
        )
        
        # Check for errors in response
        if isinstance(response, dict) and response.get("error"):
            error_msg = f"Storage upload error: {response['error']}"
            logger.error(error_msg)
            return None, error_msg
        
        logger.info(f"Successfully uploaded file to storage: {path}")
        return path, None
    
    except Exception as e:
        error_msg = f"Failed to upload file to storage: {str(e)}"
        logger.error(error_msg, exc_info=True)
        return None, error_msg


def insert_into_table(table_name: str, document_data: Dict[str, Any]) -> Tuple[bool, Optional[str], Optional[str], Optional[Dict[str, Any]]]:
    """
    Insert a row into the specified table.
    
    Args:
        document_data: Dictionary containing document fields:
            - owner_employee_id (str): UUID of the employee who owns the document
            - content (str): Extracted text content
            - content_structured (dict, optional): Structured content (for Excel files)
            - uploaded_by (str): Name of the employee who uploaded the document
            - title (str): Document title
            - ai_summary (str): AI-generated summary
            - file_url (str): Storage path of the uploaded file
            - created_at (str): ISO timestamp of creation
    
    Returns:
        Tuple of (success, document_id, error, inserted_data):
        - success (bool): Whether the insert was successful
        - document_id (str, optional): The ID of the inserted document if successful
        - error (str, optional): Error message if insert failed
        - inserted_data (dict, optional): The inserted row data if successful
    """
    try:
        supabase = get_supabase_client()
        
        # Add created_at if not provided
        if "created_at" not in document_data:
            document_data["created_at"] = datetime.now(timezone.utc).isoformat()
        
        logger.info(f"Inserting document into 'documents_1' table with fields: {list(document_data.keys())}")
        
        # Insert the row
        response = supabase.table("documents_1").insert(document_data).execute()
        
        # Check for errors
        if hasattr(response, 'error') and response.error:
            error_msg = f"Database insert error: {response.error}"
            logger.error(error_msg)
            return False, None, error_msg, None
        
        # Extract inserted data
        if hasattr(response, 'data') and response.data:
            inserted_row = response.data[0] if isinstance(response.data, list) else response.data
            inserted_id = inserted_row.get("id") if isinstance(inserted_row, dict) else None
            
            logger.info(f"Successfully inserted document with id: {inserted_id}")
            return True, inserted_id, None, inserted_row
        else:
            # No data returned but no error - might be successful but empty response
            logger.warning("Insert into 'documents_1' completed but no data returned")
            return True, None, None, None
    
    except Exception as e:
        error_msg = f"Failed to insert document: {str(e)}"
        logger.error(error_msg, exc_info=True)
        return False, None, error_msg, None

File: backend_js/src/services/database_operations.ts
/**
 * Database and file processing operations.
 *
 * This module contains operations for:
 * - File content extraction (PDF, text, Excel)
 * - Employee UUID lookup from database
 * - File storage uploads
 * - Document insertion into database
 *
 * Maps from Python database_operations.py
 */

import winston from "winston";
import {
  getSupabaseClient,
  guessContentType,
  makeStoragePath,
  BUCKET,
} from "./helpers.js";
import { readPdf, readTextFile, readExcel } from "./file_readers.js";

// Use Winston's default logger (configured in logging_config.ts)
const logger = winston;

/**
 * Read and extract content from a file based on its extension.
 *
 * Maps from Python read_file function
 */
export async function readFile(
  filename: string,
  fileBytes: Buffer
): Promise<{
  contentText: string;
  contentStructured: any | null;
  warnings: string[];
}> {
  const lower = filename.toLowerCase();

  let response: {
    kind: string;
    contentText: string;
    contentStructured: any | null;
    warnings: string[];
  };

  if (lower.endsWith(".pdf")) {
    response = await readPdf(fileBytes);
  } else if (lower.endsWith(".txt") || lower.endsWith(".md")) {
    response = readTextFile(fileBytes);
  } else if (
    lower.endsWith(".xlsx") ||
    lower.endsWith(".xlsm") ||
    lower.endsWith(".xls")
  ) {
    response = readExcel(fileBytes);
  } else {
    response = {
      kind: "unknown",
      contentText: "",
      contentStructured: null,
      warnings: ["File extension not recognized"],
    };
  }

  return {
    contentText: response.contentText || "",
    contentStructured: response.contentStructured || null,
    warnings: response.warnings || [],
  };
}

/**
 * Query the employees table to get the UUID of an employee by their employee_id.
 *
 * Maps from Python get_employee_uuid function
 */
export async function getEmployeeUuid(
  employeeId: string
): Promise<{ uuid: string | null; error: string | null }> {
  try {
    const supabase = getSupabaseClient();

    // Normalize employee_id by removing dashes (backend stores as 'EMP001', frontend may send 'EMP-001')
    const normalizedEmployeeId = employeeId.replace(/-/g, "");

    logger.info(
      `Querying employees table for employee_id: ${employeeId} (normalized: ${normalizedEmployeeId})`
    );

    // Query employees table where employee_id matches
    const { data, error } = await supabase
      .from("employees")
      .select("id")
      .eq("employee_id", normalizedEmployeeId)
      .single();

    // Check for errors
    if (error) {
      const errorMsg = `Database query error: ${error.message}`;
      logger.error(errorMsg);
      return { uuid: null, error: errorMsg };
    }

    // Extract UUID from response
    if (data && data.id) {
      const employeeUuid = data.id;
      logger.info(
        `Found employee UUID: ${employeeUuid} for employee_id: ${employeeId}`
      );
      return { uuid: String(employeeUuid), error: null };
    }

    // Employee not found
    const errorMsg = `Employee with employee_id '${employeeId}' not found`;
    logger.warn(errorMsg);
    return { uuid: null, error: errorMsg };
  } catch (error: any) {
    const errorMsg = `Failed to query employee UUID: ${error.message}`;
    logger.error(errorMsg);
    return { uuid: null, error: errorMsg };
  }
}

/**
 * Upload a file to a PRIVATE Supabase Storage bucket.
 *
 * Maps from Python upload_to_storage function
 */
export async function uploadToStorage(
  employeeId: string,
  filename: string,
  fileBytes: Buffer
): Promise<{ filePath: string | null; error: string | null }> {
  try {
    const supabase = getSupabaseClient();

    // Generate storage path (normalizes employee_id internally)
    const path = makeStoragePath(employeeId, filename);

    // Guess content type from filename
    const contentType = guessContentType(filename);

    logger.info(
      `Uploading file to storage: bucket=${BUCKET}, path=${path}, content_type=${contentType}`
    );

    // Upload to Supabase Storage
    const { error } = await supabase.storage
      .from(BUCKET)
      .upload(path, fileBytes, {
        contentType: contentType,
        upsert: true,
      });

    // Check for errors
    if (error) {
      const errorMsg = `Storage upload error: ${error.message}`;
      logger.error(errorMsg);
      return { filePath: null, error: errorMsg };
    }

    logger.info(`Successfully uploaded file to storage: ${path}`);
    return { filePath: path, error: null };
  } catch (error: any) {
    const errorMsg = `Failed to upload file to storage: ${error.message}`;
    logger.error(errorMsg);
    return { filePath: null, error: errorMsg };
  }
}

/**
 * Insert a row into the specified table.
 *
 * Maps from Python insert_into_table function
 */
export async function insertIntoTable(
  tableName: string,
  documentData: Record<string, any>
): Promise<{
  success: boolean;
  documentId: string | null;
  error: string | null;
  insertedData: any | null;
}> {
  try {
    const supabase = getSupabaseClient();

    // Add created_at if not provided
    if (!documentData.created_at) {
      documentData.created_at = new Date().toISOString();
    }

    logger.info(
      `Inserting document into '${tableName}' table with fields: ${Object.keys(documentData).join(", ")}`
    );

    // Insert the row
    const { data, error } = await supabase
      .from(tableName)
      .insert(documentData)
      .select()
      .single();

    // Check for errors
    if (error) {
      const errorMsg = `Database insert error: ${error.message}`;
      logger.error(errorMsg);
      return { success: false, documentId: null, error: errorMsg, insertedData: null };
    }

    // Extract inserted data
    if (data) {
      const insertedId = data.id || null;

      logger.info(
        `Successfully inserted document with id: ${insertedId}`
      );
      return {
        success: true,
        documentId: insertedId ? String(insertedId) : null,
        error: null,
        insertedData: data,
      };
    } else {
      // No data returned but no error - might be successful but empty response
      logger.warn(
        `Insert into '${tableName}' completed but no data returned`
      );
      return { success: true, documentId: null, error: null, insertedData: null };
    }
  } catch (error: any) {
    const errorMsg = `Failed to insert document: ${error.message}`;
    logger.error(errorMsg);
    return { success: false, documentId: null, error: errorMsg, insertedData: null };
  }
}

